Define Docker's deprecation policy

duglin opened the pull request:
Right now Docker's development is a bit constrained by the requirement to be backwards compatible.
Normally minor releases (e.g. 1.2 -> 1.3) are required to be backwards compatible, and that makes
a lot of sense for stable products that have a regular major release update schedule. 

However, Docker is still (relatively) new and as such some of the original design decisions are
no longer correct, but we are forced to support them for, what feels like, forever.  In addition,
there are cases where we decline to offer a better alternative because we would then feel 
compelled to support both forever - and usually having more than one way to do something is bad.

We have claimed to deprecate certian features in the past but it seems rare that we actually
follow through and remove support for them.

This PR attempts to formalize Docker's deprecation policy by doing a few things:
1 - makes it clear to our users what to expect with respect to backwards compatibility. See doc changes.
2 - gives us the freedom to removed deprecated features w/o having to wait for a major release,
    which has yet to happen (aside from v1.0)

There are several steps to this PR:
1 - decide if we want to even allow minor release removal of features at all
2 - if so, then decide what the proper timeframe is for keeping deprecrated features.
    This PR offers up a proposal of _2_ minor releases.
3 - pick a feature to use as a guinea pig. For now this PR has picked on `docker inspect <image>`
    just to show what the documention would look like. If we make it to step 3 in this process
    I suspect we may pick another feature.  But we can decide that later.

Signed-off-by: Doug Davis dug@us.ibm.com

Commit:
Define Docker's deprecation policy
Announce the cmd line options that are targetted for removal

Signed-off-by: Doug Davis <dug@us.ibm.com>

thaJeztah commented:
:100: thanks so much for initiating the discussion on this, @duglin, I think it's much needed. So far, docker has moved *fast*, and that's great (and needed to be able to achieve where the project is now). Docker is now widely used in production; people depend on it, and that requires stability and a clear deprecation warning well in advance (both for features and change in behaviour), giving users time to adapt.

It's less "fun" than developing new features, but equally (if not more) important for the project.

(Just my 0.02c)

duglin commented:
ping @shykes @crosbymichael

ewindisch commented:
:+1:

icecrime commented:
+1 on the idea and write-up, -1 on this specific example

crosbymichael commented on docs/misc/release-notes.md:
```diff
@@ -59,6 +59,16 @@ issues. You might have to flush your cookies if it doesn't work right away.
 For more information, see the [Docker forum
 post](https://forums.docker.com/t/new-safari-in-yosemite-issue/300).
 
+### Deprecated Features
```
you trying to sneak this in on us?

duglin replied:
LOL no it was just an example.  My number 3 even said we need to find the right example.  Any suggestions?  Thinking about recent PRs, what about dropping "docker inspect" w/o the "--type" flag?  Or drop using it for images w/o "--type image"?

crosbymichael commented on docs/misc/index.md:
```diff
@@ -99,6 +99,21 @@ implementation, check out the [Docker User Guide](/userguide/).
 A summary of the changes in each release in the current series can now be found
 on the separate [Release Notes page](/release-notes/)
 
+## Feature Deprecation Policy
```
+1

duglin commented:
wrt a feature we want to deprecate... what about support for multiple FROMs in a Dockerfile?  https://github.com/docker/docker/issues/13026    
If not, does any one have any other proposed features we should remove?

Will work on rebasing things tomorrow... the docs moves quite a bit so I need to find the new spots.

crosbymichael commented:
I cannot think of anything as an example.  I would probably make up a fake feature and be clear in the doc that it is an example. 

I would also add something to the effect of the release where the deprecation notice was first made and the release number where there feature will be removed.  Just so people can plan and don't have to do the math themself.

duglin commented:
We have quite a few CLI flags that are marked as deprecated - perhaps we could use those?

thaJeztah commented:
`docker -d -r / --restart` sounds like a good candidate, it's been replaced by restart policies in 1.2, but (just checked) it's still there.

thaJeztah commented:
Perhaps the `docker -d` flag itself; after the `docker daemon` command (https://github.com/docker/docker/pull/13771) has been implemented, it'd make sense to think about deprecating and removing that

duglin commented:
@icecrime @crosbymichael ok see if you like the new docs (picked on feature) better.

duglin commented:
ping @moxiegirl to see if I have the docs in the right spots

moxiegirl commented:
looking now

moxiegirl commented:
This is great! Thanks @duglin for starting this discussion.  Some questions I had in looking at this.

1)
Does this policy  (two releases) work for Commercially Supported Releases; I've been told by the CSR team to expect to support docs back to 1.6 for up to two years.  

2)
This solution looks purely manual and purely "doc" based to me.  That seems error prone, a doc management overload on a very small doc team, and won't scale so well long term.  Can we add/Do we have something in our CI that allows us to label a PR as "deprecated" and then later, "removal" and then auto-updating DEPRECATED and REMOVED lists.

3) 
When a user uses a command line, API, or other function/feature that is deprecated do they get a warning in some way.  In the build output? In the logs? On the command line. For example, Rails and Node.js look like they have facilities in their code for creating deprecation warnings. Mongodb appears to be using an automated tool (we used one at Atlassian) for their API deprecation list.

4)
I don't see other Docker project owners responding on this PR.  As much as possible, we should be uniform in our approach to deprecation across all Docker projects. That way users and contributors can look to common elements in each project to find deprecation. 

5) Have we looked at how other projects are handling this common type issue. Anything good out there?
http://guides.rubyonrails.org/4_2_release_notes.html#railties-deprecations
https://docs.puppetlabs.com/puppet/3.7/reference/deprecated_command.html
http://api.mongodb.org/java/current/deprecated-list.html

### About the approach in this PR
My general recommendations/comments/thoughts are:
- Deprecation Policy should be short and  repeated in each release note. (I like the policy -- short and sweet)
- **Deprecated** and **Removed** are typically headings  in release notes; we should do that however the lists are compiled (hopefully not manually)
- Having notices in the release notes will provide timestamp/release version info automatically for each item

thaJeztah commented:
Apart from a clear deprecation policy, a migration guide between versions could be helpful (although this will not be something that can be automated). For example, this is how the CakePHP project describes migrations between versions; [CakePHP 2.1 migration guide](http://book.cakephp.org/2.0/en/appendices/2-1-migration-guide.html)

The PHP manual uses a changelog *per function*, see for example the [json_encode() documentation](http://php.net/manual/en/function.json-encode.php) similarly, the Mozilla Developer Network (MDN) docs keeps a compatibility list per page, for example, the [HTML `<blink>` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/blink)

I think that having a changelog/deprecation warnings per page could make it easier to maintain (and easier for users to find). If the right format is used, we could automatically collect those change logs / deprecation warnings and generate a global changelog.

duglin commented:
On Jul 10, 2015, at 10:51 AM, moxiegirl <notifications@github.com> wrote:
> This is great! Thanks @duglin <https://github.com/duglin> for starting this discussion. Some questions I had in looking at this.
> 1)
> Does this policy (two releases) work for Commercially Supported Releases; I've been told by the CSR team to expect to support docs back to 1.6 for up to two years.
> 
“support” and “backwards compatibility” are different. The CSR team, actually each company’s CSR-like team, will need to decide how far back they’ll go w.r.t. supporting older Docker releases.  The Docker community will need to decide how to manage backwards compatibility.  IMO, we can’t keep broken/old things around forever and waiting for a major release is just way too long for a fast moving project like Docker - so a default position of 2 releases seems like a good starting point, but we’ll need to look at each feature to see what’s best for that area of the code.  One nice thing though, is that the minute we put an item no the chopping block it should cause the people who care about it to speak up - and if they have good reasons for us to change our minds then we can do so. Forces the discussion. 
> 2)
> 
> This solution looks purely manual and purely "doc" based to me. That seems error prone, a doc management overload on a very small doc team, and won't scale so well long term. Can we add/Do we have something in our CI that allows us to label a PR as "deprecated" and then later, "removal" and then auto-updating DEPRECATED and REMOVED lists.
> 
I like the idea of tagging PRs with a “deprecated” type of label so that we know whether there should be a corresponding item in the “deprecated features list”.  I’m not sure how to get the automated stuff to work though since the description of the feature, why its being removed, what its replacement is, etc... is pretty manual.  Although, any PR with a “deprecated” label should modify the “deprecated.md” file in the docs - and that should be up to the dev to do (at least for a first pass), not the docs folks.
> 3) 
> When a user uses a command line, API, or other function/feature that is deprecated do they get a warning in some way. In the build output? In the logs? On the command line. For example, Rails and Node.js look like they have facilities in their code for creating deprecation warnings. Mongodb appears to be using an automated tool (we used one at Atlassian) for their API deprecation list.
> 
I think it depends on the feature.  In the case of the first example (cmd line opts) they do get a warning and I think in general that would be a good thing to do.  But, I suspect that there will be some features that won’t have a nice place to spit out warnings but I agree, it should be our goal to try if we can.
> 4)
> I don't see other Docker project owners responding on this PR. As much as possible, we should be uniform in our approach to deprecation across all Docker projects. That way users and contributors can look to common elements in each project to find deprecation.
> 
ping @docker/maintainers <https://github.com/orgs/docker/teams/maintainers>
I’ll send a note to the maintainer’s distribution list too.
> 5) Have we looked at how other projects are handling this common type issue. Anything good out there?
> http://guides.rubyonrails.org/4_2_release_notes.html#railties-deprecations <http://guides.rubyonrails.org/4_2_release_notes.html#railties-deprecations>
> https://docs.puppetlabs.com/puppet/3.7/reference/deprecated_command.html <https://docs.puppetlabs.com/puppet/3.7/reference/deprecated_command.html>
> http://api.mongodb.org/java/current/deprecated-list.html <http://api.mongodb.org/java/current/deprecated-list.html>These seem to do something similar. I just modified the guinea pig feature to include a “Replacement” section so that we have a well defined spot to include text telling people what they should use instead - if there is anything.

ping @icecrime @crosbymichael to make sure they’re ok with this new text.
> About the approach in this PR
> 
> My general recommendations/comments/thoughts are:
> 
> Deprecation Policy should be short and repeated in each release note. (I like the policy -- short and sweet)
I did try to include something in the release notes doc - do you think I need more?
> Deprecated and Removed are typically headings in release notes; we should do that however the lists are compiled (hopefully not manually)
I didn’t create a “Removed” page yet. Let me work on that.
> Having notices in the release notes will provide timestamp/release version info automatically for each item

Not sure what edits you want me to make based on this comment.

duglin commented:
@moxiegirl I modified the Release notes to include a list of "Removed features" - granted it empty for now :-)

calavera commented:
two releases sounds like a rather short period of time. Lately, I've seen people updating from docker 1.3.3 and docker 1.4 to docker 1.7. I think three or four releases sounds more adequate to give people time enough to prepare their migrations.

icecrime commented:
LGTM

Regarding the "grace period", I think two releases is good enough. Just as an FYI, I've been playing with Elastic Search lately and I like how obvious things are in their documentations ([example](https://www.elastic.co/guide/en/elasticsearch/reference/1.6/search-aggregations-bucket-datehistogram-aggregation.html#_time_zone)).

duglin commented:
@jfrazelle how do we go about adding a "deprecated-feature" label for PRs?

jfrazelle commented:
um it is a github thing, you click labels and add new or something

On Tue, Jul 14, 2015 at 3:10 PM, Doug Davis <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> how do we go about adding a
> "deprecated-feature" label for PRs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13900#issuecomment-121407180>.
>

thaJeztah commented:
Label should probably be incorporated in https://github.com/docker/docker/pull/14445 (and preferably be named and used the same in all repos)

duglin commented:
@jfrazelle I must not have access to do that   :-(

jfrazelle commented:
oh lol hahaha

On Tue, Jul 14, 2015 at 3:52 PM, Doug Davis <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> I must not have access to do
> that :-(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13900#issuecomment-121415535>.
>

SvenDowideit commented on docs/misc/deprecated.md:
```diff
@@ -0,0 +1,64 @@
+The following list of features are deprecated.
+
+### Old Command Line Options
+**Target For Removal In Release: v1.10**
```
Can we add 

`(Deprecated from Release: v1.8.0)` (and possibly a date, or link to that version's release note)?

duglin replied:
Done.

SvenDowideit commented on docs/misc/deprecated.md:
```diff
@@ -0,0 +1,64 @@
+**Replacement:**
+
+Many of the deprecated options have a double-dash variant that should be
+used instead. There is no replacement option available for those that do not.
```
Can this document also list or link to the API changes - this might be something that could be automated, but we've surprised users a number of times, so over-documentation is probably worth it.

jamtur01 replied:
I'd make this much clearer. The first list of single-dash variants are deprecated and replaced with double-dash variants. The second list are just deprecated. I'd remove the **Replacement** section altogether. Perhaps:

```The following single-dash (`-opt`) variant of certain command line options are deprecated and replaced with double-dash options (`--opt`):```

then

```The following single-dash options are deprecated and have no replacement:```

duglin replied:
@jamtur01 made your suggested edits.
@SvenDowideit I think we could but I'm not sure if a doc with that list exists - does it?

SvenDowideit commented:
very nice LGTM.

jamtur01 commented on docs/misc/deprecated.md:
```diff
@@ -0,0 +1,64 @@
+**Target For Removal In Release: v1.10**
+
+The following single-dash (`-opt`) variant of certain command line options
+are deprecrated:
```
deprecated

duglin replied:
fixed

jamtur01 commented on docs/misc/release-notes.md:
```diff
@@ -9,6 +9,17 @@ parent = "smn_release_notes"
+# Deprecated Features
+
+To see the complete list of deprecated features please see the
+[Deprecated Features](deprecrated) page.
```
deprecated

duglin replied:
fixed

jamtur01 commented on docs/misc/release-notes.md:
```diff
@@ -9,6 +9,17 @@ parent = "smn_release_notes"
+
+# Removed Features
+
+The following features have been removed this release:
```
```The following features have been removed in this release:```

duglin replied:
fixed

jamtur01 commented:
Needs some language fine tuning as per comments I think but overall concept +1.

jfrazelle commented:
where we at with this ;)

duglin commented:
@jfrazelle   grrr sorry - keep getting distracted. Let me try to address the comment now....

duglin commented:
ok - I got I got all edits.
ping @jamtur01 @SvenDowideit @moxiegirl 
ping @tiborvass the "docker daemon" stuff should be mentioned in here if both make it into 1.8

jamtur01 commented:
Docs LGTM

thaJeztah commented:
LGTM, thanks!

duglin commented:
thanks all!

tiborvass commented on docs/misc/deprecated.md:
```diff
@@ -0,0 +1,58 @@
+    docker search -trusted
+    docker tag -force
+
+The following single-dash options are deprecated and have no replacement:
```
@duglin didn't you mean `double-dash` ?

duglin replied:
oh carp!  thanks for noticing.
pr #14886 to fix it.

thaJeztah commented:
Note that we may change the policy to 3 release cycles; see https://github.com/docker/docker/pull/24572, and https://github.com/docker/docker/issues/24534
